addons/l10n_eg_edi_eta/models/account_edi_format.py

```diff
@@ -0,0 +1,386 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+
+import json
+import logging
+import requests
+from werkzeug.urls import url_quote
+from base64 import b64encode
+
+from odoo import api, models, _
+from odoo.tools.float_utils import float_repr, float_round
+
+_logger = logging.getLogger(__name__)
+
+
+ETA_DOMAINS = {
+    'preproduction': 'https://api.preprod.invoicing.eta.gov.eg',
+    'production': 'https://api.invoicing.eta.gov.eg',
+    'token.preproduction': 'https://id.preprod.eta.gov.eg',
+    'token.production': 'https://id.eta.gov.eg',
+}
+
+
+class AccountEdiFormat(models.Model):
+    _inherit = 'account.edi.format'
+
+    @api.model
+    def _l10n_eg_get_eta_api_domain(self, production_enviroment=False):
+        return production_enviroment and ETA_DOMAINS['production'] or ETA_DOMAINS['preproduction']
+
+    @api.model
+    def _l10n_eg_get_eta_token_domain(self, production_enviroment=False):
+        return production_enviroment and ETA_DOMAINS['token.production'] or ETA_DOMAINS['token.preproduction']
+
+    @api.model
+    def _l10n_eg_eta_connect_to_server(self, request_data, request_url, method, access_token=False, production_enviroment=False):
+        api_domain = access_token and self._l10n_eg_get_eta_token_domain(production_enviroment) or self._l10n_eg_get_eta_api_domain(production_enviroment)
+        request_url = request_url % api_domain
+        try:
+            request_response = requests.request(method, request_url, data=request_data.get('body'), headers=request_data.get('header'), timeout=(5, 10))
+        except (ValueError, requests.exceptions.ConnectionError, requests.exceptions.MissingSchema, requests.exceptions.Timeout, requests.exceptions.HTTPError) as ex:
+            return {
+                'error': str(ex),
+                'blocking_level': 'warning'
+            }
+        if not request_response.ok:
+            response_data = request_response.json()
+            if isinstance(response_data, dict) and response_data.get('error'):
+                return {
+                    'error': response_data.get('error', _('Unknown error')),
+                    'blocking_level': 'error'
+                }
+            return {
+                'error': request_response.reason,
+                'blocking_level': 'error'
+            }
+        return {'response': request_response}
+
+    @api.model
+    def _l10n_eg_edi_round(self, amount):
+        return float(float_repr(float_round(amount, precision_digits=5), precision_digits=5))
```

Comment:
@qdp-odoo is right, this demonstrates a clear misunderstanding of rounding. By casting back into a float you're **undoing** what `float_round` did to fix rounding errors.

Example: 3.175 has no exact representation, and Python will approach it from below, causing possible rounding errors. So `float_round` will fix that by adding an epsilon to tip the rounding in the right direction:

```
>>> "%.30f" % 3.175
'3.174999999999999822364316059975'
>>> round(3.175, 2)
3.17  # << rounding error, 3.175 should become 3.18

>>> f = float_round(3.175, precision_digits=5)
>>> round(f, 2)
3.18  # << correct
```

Of course `float_round` may look like it produces display errors with extra decimals, but that is only a *local representation* while you debug. When stored in the database the value will be converted by `float_repr` automatically to be exact, to the right precision:

```
>>> f
3.1750000000000003  # looks weird, but is just a local repr

>>> float_repr(f, precision_digits=5)
'3.17500'   # << as stored in database by float/decimal field
```

Here in `_l10n_eg_edi_round` you're casting the float_repr back into a float, that means you're undoing all the work of rounding, and you're creating a float that can't be rounded correctly. That's useless, you could just `return amount` and get the same result.
```
>>> round(_l10n_eg_edi_round(3.175), 2)
3.17    # << can't be rounded correctly
```

---

SUMMARY: don't try to outsmart the framework, use the tools as they are supposed to be used:
- use `float_round()` or `currency.round()` when you're mixing different precisions (e.g. UoM vs currencies) and you want to be sure of the rounding, or if you're not storing in a field with a defined precision. In most cases `Float` and `Monetary` fields will round automatically, and you probably don't need to [^1]
- use `float_repr()` when you want to output a rounded float value outside of the database, e.g. in a file. Most float/monerary widgets already do it for you in QWeb/Web Client.
- never cast a `float_repr` result back into a float, you're re-creating the rounding errors

[^1]: https://github.com/odoo/odoo/blob/master/odoo/fields.py#L1490

Reply:
If the point is to encode float values for `json.dumps`:
- the naming should clearly reflect this, unlike the current method name
- the current strategy is incorrect, you could still end up with an incorrectly represented float. You need a custom class whose `repr()` provides the result from `float_repr()` directly
- isn't there already a framework solution for this problem? if not, shouldn't we introduce it, instead of duplicating this?